### PR TITLE
[INLONG-6584][Sort] Add read phase metric and table level metric for MySQL

### LINKED ISSUE #6584: [Improve][Sort] Add read phase metric and table level metric  for MySQL
### Description

_No response_

### InLong Component

InLong Sort

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by chestnut-c**
### Prepare a Pull Request

- [INLONG-6584][Sort] Add read phase metric and table level metric for MySQL

- Fixes #6584

### Motivation

   The background is that when the entire database is migrated, the business needs to count the indicators of table granularity. Dynamically generate table granularity MetricGroup according to the schema of data records; internally maintain MetricGroup container to realize table granularity reporting indicators.

### Modifications

Add read phase metric and table level metric for MySQL

### Verifying this change

*(Please pick either of the following options)*

- [ ] This change is a trivial rework/code cleanup without any test coverage.

- [ ] This change is already covered by existing tests, such as:
  *(please describe tests)*

- [ ] This change added tests and can be verified as follows:

  *(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation

### CODE REVIEW

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java [hunk 1]
```diff
@@ -73,11 +76,44 @@ public final class Constants {
...
+    /**
+     * It is used for inlong.metric
+     */
+    public static final Long SNAPSHOT_PHASE = 1L;
+
+    /**
+     * It is used for inlong.metric
+     */
+    public static final Long INCREASE_PHASE = 2L;
+
```
Comment (gong) [inline inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java:116]: `It is used for inlong.metric` should be modified.

**Review by gong — APPROVED** [review]

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/MetricType.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * connector metric type enum
+ */
+public enum MetricType {
+
+    /**
+     * node level metric
+     */
+    NODE,
```
Comment (thesumery) [inline inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/MetricType.java:29]: node level why not add streamId groupId nodeId label?

Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/MetricType.java:29]: > node level why not add streamId groupId nodeId label?

The node level has added streamId groupId nodeId tags.

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/SourceMetricData.java [hunk 3]
```diff
@@ -222,18 +389,119 @@ public void outputMetrics(long rowCountSize, long rowDataSize) {
...
                 + '}';
     }
+
+    /**
+     * Source Record Schema Info
+     */
+    public static class SourceRecordSchemaInfo {
+
+        @NonNull
+        private String databaseName;
```
[the remark this answers is not in the document]
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/SourceMetricData.java:433]: ok，I will add the form of kafka topic, although it is not used at present.

**Review by EMsnap — APPROVED** [review]

**Comment by chestnut-c** [comment]
> Please explain in detail the requirements background and design ideas？


 hi , The background is that when the entire database is migrated, the business needs to count the indicators of table granularity. Dynamically generate table granularity MetricGroup according to the schema of data records; internally maintain MetricGroup container to realize table granularity reporting indicators.

**Review by healchow — APPROVED** [review]

**Comment by yunqingmoswu** [comment]
> > Please explain in detail the requirements background and design ideas？
> 
> Hi , The background is that when the entire database is migrated, the business needs to count the indicators of table granularity. Dynamically generate table granularity MetricGroup according to the schema of data records; internally maintain MetricGroup container to realize table granularity reporting indicators.

You can  add this to the describe of pr.

**Review by yunqingmoswu — COMMENTED** [review]
It is recommended to extend the option 'inlong.metric.labels' instead of adding metric type.
For example you can format as 'groupId=xxx&streamId=xxx&nodeId=xxx&database=${database}&table=${table}', in this format the '${xxx}'  represents a variable, and the specific supported variables are determined by the connector.

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java [hunk 4]
```diff
@@ -57,6 +57,9 @@ public final class Constants {
     public static final String NUM_BYTES_IN_PER_SECOND = "numBytesInPerSecond";
 
     public static final String NUM_RECORDS_IN_PER_SECOND = "numRecordsInPerSecond";
+
+    public static final String READ_PHASE = "readPhase";
```
[the remark this answers is not in the document]
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java:61]: > Please add some comment for it,simply looking at this variable does not know the specific meaning?

The whole database migration stage distinguishes between snapshot and incremental data, business indicators need to know the current stage。

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java [hunk 5]
```diff
@@ -73,11 +76,48 @@ public final class Constants {
      * Node id used in inlong metric
      */
     public static final String NODE_ID = "nodeId";
+    /**
+     * Database Name used in inlong metric
+     */
+    public static final String DATABASE_NAME = "database";
```
[the remark this answers is not in the document]
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java:82]: > These constants are not considered general concepts, it is recommended to consider a more general design？

Specific tags are listed for flexible business use.

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/MetricType.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.apache.inlong.sort.base;
+
+/**
+ * connector metric type enum
+ */
+public enum MetricType {
```
[the remark this answers is not in the document]
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/MetricType.java:24]: > It seems unnecessary to define the metric type, just extend it on the option 'inlong.metric.label'.

This enumeration is to distinguish Node indicators or more fine-grained indicators. Fine-grained indicators are currently expanded on the main node indicators.

[the remark this answers is not in the document]
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/MetricType.java:24]: > It seems unnecessary to define the metric type, just extend it on the option 'inlong.metric.label'.

i will remove it.

**Comment by chestnut-c** [comment]
> > > Please explain in detail the requirements background and design ideas？
> > 
> > 
> > Hi , The background is that when the entire database is migrated, the business needs to count the indicators of table granularity. Dynamically generate table granularity MetricGroup according to the schema of data records; internally maintain MetricGroup container to realize table granularity reporting indicators.
> 
> You can add this to the describe of pr.

ok

**Comment by chestnut-c** [comment]
> It is recommended to extend the option 'inlong.metric.labels' instead of adding metric type. For example you can format as 'groupId=xxx&streamId=xxx&nodeId=xxx&database=${database}&table=${table}', in this format the '${xxx}' represents a variable, and the specific supported variables are determined by the connector.



> 'inlong.metric.labels'

If you use what you said to do it, it feels a bit redundant. First of all, you can’t directly get the label at the node level, unless you add 'inlong.metric.tableLabels'. It is better to hand over the specific sub-label to the specific connector.

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java:61]: > Please add some comment for it,simply looking at this variable does not know the specific meaning?

ok. i will  some comment.

**Comment by chestnut-c** [comment]
> > > Please explain in detail the requirements background and design ideas？
> > 
> > 
> > Hi , The background is that when the entire database is migrated, the business needs to count the indicators of table granularity. Dynamically generate table granularity MetricGroup according to the schema of data records; internally maintain MetricGroup container to realize table granularity reporting indicators.
> 
> You can add this to the describe of pr.

ok i will add it

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java [hunk 7]
```diff
@@ -73,11 +79,48 @@ public final class Constants {
...
+
+    /**
+     * It is used for metric data to spilt schema identify
+     */
+    public static final String SPILT_SEMICOLON = "\\.";
+
+    /**
+     * This indicator represents the snapshot phase during the read phase
+     */
+    public static final Long SNAPSHOT_PHASE = 1L;
```
Comment (EMsnap) [inline inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java:117]: suggest use enum for snapshot or incremental

**Review by e-mhui — APPROVED** [review]

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/SourceMetricData.java [hunk 8]
```diff
@@ -189,11 +237,126 @@ public Map<String, String> getLabels() {
...
+        identifyBuilder.append(table);
+        return identifyBuilder.toString();
+    }
+
     public void outputMetricsWithEstimate(Object o) {
         long size = o.toString().getBytes(StandardCharsets.UTF_8).length;
         outputMetrics(1, size);
     }
 
+    public void outputMetricsWithEstimate(SourceRecordSchemaInfo recordSchemaInfo, Object o) {
```
Comment (EMsnap) [inline inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/SourceMetricData.java:339]: Object data

Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/SourceMetricData.java]: > Object data

OK ,i will change the name to data.

File: inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java (hunk 7 above)
Comment (chestnut-c) [reply inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/Constants.java]: > suggest use enum for snapshot or incremental

good idea ,i will do it .

**Comment by yunqingmoswu** [comment]
> > It is recommended to extend the option 'inlong.metric.labels' instead of adding metric type. For example you can format as 'groupId=xxx&streamId=xxx&nodeId=xxx&database=${database}&table=${table}', in this format the '${xxx}' represents a variable, and the specific supported variables are determined by the connector.
> 
> > 'inlong.metric.labels'
> 
> If you use what you said to do it, it feels a bit redundant. First of all, you can’t directly get the label at the node level, unless you add 'inlong.metric.tableLabels'. It is better to hand over the specific sub-label to the specific connector.
Normally, it should be like this. The reporting of indicators based on nodeId is closed, so it is a closed loop inside sort, but this function has broken this agreement and is open. Then, general rules should be defined, and each connector can implement it according to the situation.

**Review by thesumery — APPROVED** [review]

**Pull request closed without merging**